src/createServer.js

```diff
@@ -1,11 +1,163 @@
 'use strict';
 
-// const express = require('express');
+const express = require('express');
 
 function createServer() {
-  // Use express to create a server
-  // Add a routes to the server
-  // Return the server (express app)
+  const app = express();
+
+  let users = [];
+
+  app.use(express.json());
+
+  app.get('/users', (req, res) => {
+    res.setHeader('Content-Type', 'application/json');
+    res.send(users);
+  });
+
+  app.post('/users', (req, res) => {
+    res.setHeader('Content-Type', 'application/json');
+
+    if (req.body.name) {
+      const user = {
+        id: Math.max(0, ...users.map(u => u.id)) + 1,
+        name: req.body.name,
+      };
+
+      users.push(user);
+
+      return res.status(201).send(user);
+    } else {
+      res.sendStatus(400);
+    }
+  });
+
+  app.get('/users/:userId', (req, res) => {
+    res.setHeader('Content-Type', 'application/json');
+
+    const { userId } = req.params;
+    const user = users.find(u => u.id === Number(userId));
+
+    if (user) {
+      res.send(user);
+    } else {
+      res.sendStatus(404);
+    }
+  });
+
+  app.delete('/users/:userId', (req, res) => {
+    const { userId } = req.params;
+    const user = users.find(u => u.id === Number(userId));
+
+    if (user) {
+      users = users.filter(u => u.id !== Number(userId));
+
+      res.sendStatus(204);
+    } else {
+      res.sendStatus(404);
+    }
+  });
+
+  app.patch('/users/:userId', (req, res) => {
+    res.setHeader('Content-Type', 'application/json');
+
+    const { userId } = req.params;
+    const user = users.find(u => u.id === Number(userId));
+
+    if (user) {
+      user.name = req.body.name;
+      res.send(user);
+    } else {
+      res.sendStatus(404);
+    }
+  });
+
+  // Expenses
+
+  let expenses = [];
+
+  app.get('/expenses', (req, res) => {
+    res.setHeader('Content-Type', 'application/json');
+
+    const { userId, from, to, categories } = req.query;
+    let filteredExpenses = [...expenses];
+
+    if (userId) {
+      filteredExpenses = filteredExpenses
+        .filter(e => e.userId === Number(userId));
+    }
+
+    if (from && to) {
+      filteredExpenses = filteredExpenses
+        .filter(e => e.spentAt > from && e.spentAt < to);
+    }
+
+    if (categories) {
+      filteredExpenses = filteredExpenses
+        .filter(e => categories.includes(e.category));
+    }
```

Comment:
please use condition in one `.filter` method, because now it will invoke filter a lot of times of there is a multiple filters in query